ObjectiveGit/GTNote.m

```diff
@@ -0,0 +1,110 @@
+//
+//  GTNote.m
+//  ObjectiveGitFramework
+//
+//  Created by Slava Karpenko on 16.05.16.
+//  Copyright © 2016 Wildbit LLC. All rights reserved.
+//
+
+#import "GTNote.h"
+#import "NSError+Git.h"
+#import "GTSignature.h"
+#import "GTReference.h"
+#import "GTRepository.h"
+#import "NSString+Git.h"
+#import "GTOID.h"
+
+#import "git2/errors.h"
+#import "git2/notes.h"
+
+@interface GTNote ()
+{
+	git_note *_note;
+}
+
+@end
+@implementation GTNote
+
+- (NSString *)description {
+	return [NSString stringWithFormat:@"<%@: %p>", NSStringFromClass([self class]), self];
+}
+
+#pragma mark API
+
+- (void)dealloc {
+	if (_note != NULL) {
+		git_note_free(_note);
+	}
+}
+
+- (git_note *)git_note {
+	return _note;
+}
+
+- (NSString *)note {
+	return @(git_note_message(self.git_note));
+}
+
+- (GTSignature *)author {
+	return [[GTSignature alloc] initWithGitSignature:git_note_author(self.git_note)];
+}
+
+- (GTSignature *)committer {
+	return [[GTSignature alloc] initWithGitSignature:git_note_committer(self.git_note)];
+}
+
+- (GTOID *)targetOID {
+	return [GTOID oidWithGitOid:git_note_id(self.git_note)];
+}
+
+- (instancetype)initWithTargetOID:(GTOID *)oid repository:(GTRepository *)repository referenceName:(NSString *)referenceName error:(NSError **)error {
+	self = [super init];
+	if (self == nil) return nil;
+	
+	int gitErr = git_note_read(&_note, repository.git_repository, referenceName.UTF8String, oid.git_oid);
+	
+	if (gitErr != GIT_OK && error != NULL) *error = [NSError git_errorFor:gitErr description:@"Failed to read a note."];
+	
+	if (gitErr != GIT_OK) return nil;
+	
+	return self;
+}
+
+- (instancetype)initWithTargetGitOID:(git_oid *)oid repository:(git_repository *)repository referenceName:(const char *)referenceName {
+	self = [super init];
+	if (self == nil) return nil;
+	
+	int gitErr = git_note_read(&_note, repository, referenceName, oid);
+	
+	if (gitErr != GIT_OK) return nil;
+	
+	return self;
+}
+
+- (instancetype)init {
+	NSAssert(NO, @"Call to an unavailable initializer.");
+	return nil;
+}
+
++ (NSString *)defaultReferenceNameForRepository:(GTRepository *)repository error:(NSError **)error {
+	NSString *noteRef = nil;
+	
+	git_buf default_ref_name = { 0 };
+	int gitErr = git_note_default_ref(&default_ref_name, repository.git_repository);
+	if (gitErr != GIT_OK) {
+		if (error != NULL) *error = [NSError git_errorFor:gitErr description:@"Unable to get default git notes reference name"];
+		
+		return nil;
+	}
+	
+	if (default_ref_name.ptr != NULL) {
+		noteRef = @(default_ref_name.ptr);
+	} else {
+		if (error != NULL) *error = [NSError git_errorFor:GIT_ERROR description:@"Unable to get default git notes reference name"];
```

Comment:
But if gitErr is not GIT_OK, then it would return that error earlier. If we got to that point when `git_note_default_ref` returned `GIT_OK`, but `default_ref_name.ptr` is `NULL` for some reason, then I am generating a generic `GIT_ERROR`. I know in reality it shouldn't happen, but I am being just extra cautious here.